SOURCES/etc/xapi.d/plugins/updater.py

```diff
@@ -106,44 +106,30 @@ def display_package(p):
     return {'name': p.name, 'version': p.version, 'release': p.release, 'description': p.summary,
             'changelog': changelog, 'url': p.url, 'size': p.size, 'license': p.license}
 
-def build_repo_list(additional_repos):
+def build_repo_list(enabled_repos, additional_repos):
     repos = list(DEFAULT_REPOS)
     if additional_repos:
         repos += [x.strip() for x in additional_repos.split(',')]
-    return repos
+    return filter(lambda repo: repo in [x.id for x in enabled_repos], repos)
+
+def install_helper(session, args, mode):
+    assert mode == 'install' or mode == 'update'
 
-@error_wrapped
-@operationlock()
-def check_update(session, args):
-    repos = build_repo_list(args.get('repos'))
     yum_instance = yum.YumBase()
     yum_instance.preconf.debuglevel = 0
-    yum_instance.preconf.plugins = True
-    yum_instance.repos.disableRepo('*')
-    yum_instance.repos.enableRepo(','.join(repos))
-    packages = yum_instance.doPackageLists(pkgnarrow='updates')
-    del yum_instance.ts
-    yum_instance.initActionTs()  # make a new, blank ts to populate
-    yum_instance.populateTs(keepold=0)
-    yum_instance.ts.check()  # required for ordering
-    yum_instance.ts.order()
-    # Python 2&3 compatible code
-    return json.dumps(list(map(display_package, packages)))
+    yum_instance.preconf.plugins = False
+    repos = build_repo_list(yum_instance.repos.listEnabled(), args.get('repos'))
 
-@error_wrapped
-@operationlock(timeout=10)
-def update(session, args):
-    repos = build_repo_list(args.get('repos'))
     task = None
     res = None
     error = None
     try:
         host = session.xenapi.session.get_this_host(session.handle)
         host_name = session.xenapi.host.get_name_label(host)
         host_uuid = session.xenapi.host.get_uuid(host)
-        task = session.xenapi.task.create('Update host %s (%s)' % (host_name, host_uuid), '')
+        task = session.xenapi.task.create('%s host %s (%s)' % (mode.capitalize(), host_name, host_uuid), '')
```

Comment:
When mode is install, the message becomes confusing: "Install host host_name (host_uuid)". We're not installing a host, we're installing packages on a host.
Plus, the package(s) to be installed should be listed in the message.

Example "Install package1 package2 package3 on host host_name (host_uuid)".

For updates, when there are no specified packages, then the current message is good enough, but if there's a specified package, then it could also look like the one for installation: "Update package1 package2 package3 on host host_name (host_uuid)"